File: apps/ml-python/app/moderator_dashboard.py
"""Domain 70 — Moderator Dashboard ML scorers (deterministic, explainable)."""
from __future__ import annotations
from typing import Any, Dict

from fastapi import APIRouter
from pydantic import BaseModel, Field

router = APIRouter(prefix="/moderator-dashboard", tags=["moderator-dashboard"])


class ContentIn(BaseModel):
    surface: str | None = None
    targetId: str | None = None
    reasonCode: str = "other"
    severity: str = "normal"
    reporterId: str | None = None
    evidence: list = Field(default_factory=list)
    meta: Dict[str, Any] = Field(default_factory=dict)


@router.post("/score")
def score(body: ContentIn):
    sev_map = {"low": 20, "normal": 40, "high": 70, "critical": 90}
    sev = sev_map.get(body.severity, 40)
    high_risk = {"csam", "illegal", "self_harm", "hate"}
    med_risk = {"harassment", "impersonation", "scam", "phishing"}
    boost = 25 if body.reasonCode in high_risk else 15 if body.reasonCode in med_risk else 0
    s = min(100, sev + boost)
    if s >= 80: band = "critical"
    elif s >= 60: band = "high"
    elif s >= 35: band = "elevated"
    else: band = "normal"
    return {
        "score": s, "band": band,
        "reasons": [f"reason:{body.reasonCode}", f"severity:{body.severity}"],
    }


class RiskIn(BaseModel):
    signals: Dict[str, Any] = Field(default_factory=dict)


@router.post("/risk")
def risk(body: RiskIn):
    s = body.signals or {}
    by_status = s.get("byStatus") or {}
    breached = int(s.get("slaBreached") or 0)
    incidents = int((s.get("messagingByStatus") or {}).get("pending") or 0)
    open_count = sum(int(by_status.get(k, 0)) for k in ("open", "triaging", "holding"))
    escalated = int(by_status.get("escalated") or 0)
    score_v = 10 + min(40, open_count * 2) + min(25, escalated * 8) + min(15, breached * 5) + min(10, incidents * 3)
    score_v = min(100, score_v)
    if score_v >= 80: band = "critical"
    elif score_v >= 60: band = "high"
    elif score_v >= 35: band = "elevated"
    else: band = "normal"
    return {
        "score": score_v, "band": band, "model": "deterministic-v1",
        "factors": {
            "open": open_count, "escalated": escalated,
            "slaBreached": breached, "pendingIncidents": incidents,
        },
    }

File: packages/sdk/src/moderator-dashboard.ts
/** Typed SDK for Domain 70 — Moderator Dashboard. */

export type ModRole = 'viewer' | 'moderator' | 'senior_moderator' | 'trust_safety_admin';
export type ItemStatus = 'open' | 'triaging' | 'holding' | 'escalated' | 'actioned' | 'dismissed' | 'closed';
export type ItemQueue  = 'triage' | 'review' | 'escalation' | 'messaging_incident' | 'closed';
export type Surface    =
  | 'post' | 'profile' | 'message' | 'media' | 'comment' | 'review'
  | 'project' | 'gig' | 'service' | 'job' | 'dm_thread';
export type ReasonCode =
  | 'spam' | 'harassment' | 'hate' | 'csam' | 'illegal'
  | 'impersonation' | 'intellectual_property' | 'self_harm' | 'nsfw' | 'scam' | 'other';
export type Severity = 'low' | 'normal' | 'high' | 'critical';
export type ModAction =
  | 'warn' | 'hide' | 'remove' | 'quarantine' | 'suspend' | 'ban'
  | 'escalate_legal' | 'escalate_trust_safety' | 'dismiss' | 'restore' | 'none';
export type IncidentSignal =
  | 'keyword' | 'rate_limit' | 'phishing' | 'solicitation' | 'grooming'
  | 'self_harm' | 'threat' | 'spam' | 'user_report' | 'automation' | 'other';
export type IncidentStatus = 'pending' | 'reviewed' | 'actioned' | 'dismissed' | 'escalated';

export interface ModItem {
  id: string; reference: string; surface: Surface; target_id: string;
  reason_code: ReasonCode; reason_detail?: string | null;
  reporter_id?: string | null; evidence: Array<{ url: string; label: string }>;
  severity: Severity; ml_score: number; ml_band: 'normal' | 'elevated' | 'high' | 'critical';
  ml_reasons: string[]; status: ItemStatus; queue: ItemQueue;
  assigned_to?: string | null; sla_due_at?: string | null;
  meta: Record<string, unknown>; created_at: string; updated_at: string;
}
export interface ModActionRow {
  id: string; item_id: string; actor_id: string; action: ModAction;
  rationale: string; duration_h?: number | null; appealable: 'yes' | 'no';
  meta: Record<string, unknown>; created_at: string;
}
export interface ModEvent {
  id: string; item_id?: string | null; incident_id?: string | null;
  actor_id?: string | null; action: string;
  from_state?: string | null; to_state?: string | null;
  diff: Record<string, unknown>; created_at: string;
}
export interface MessagingIncident {
  id: string; item_id?: string | null; thread_id: string;
  participants: Array<{ userId: string; role?: string }>;
  signal: IncidentSignal; excerpt?: string | null;
  ml_score: number; ml_band: string; status: IncidentStatus;
  reviewed_by?: string | null; reviewed_at?: string | null;
  meta: Record<string, unknown>; created_at: string;
}
export interface ModMacro { id: string; slug: string; label: string; action: ModAction; template: string; meta: Record<string, unknown>; created_at: string }

export interface ModKpis {
  byStatus: Partial<Record<ItemStatus, number>>;
  byQueue: Partial<Record<ItemQueue, number>>;
  bySeverity: Partial<Record<Severity, number>>;
  slaBreached: number;
  messagingByStatus: Partial<Record<IncidentStatus, number>>;
}
export interface ModInsight { id: string; severity: 'success' | 'info' | 'warn' | 'critical'; title: string }
export interface ModRiskScore { score: number; band: 'normal' | 'elevated' | 'high' | 'critical'; model: string; factors: Record<string, number> }
export interface ModOverview {
  kpis: ModKpis;
  queues: { triage: ModItem[]; review: ModItem[]; escalation: ModItem[] };
  messagingIncidents: MessagingIncident[];
  insights: ModInsight[]; riskScore: ModRiskScore; computedAt: string;
}
export interface ModItemDetail { item: ModItem; actions: ModActionRow[]; events: ModEvent[] }
export interface ModList<T> { items: T[]; total: number; meta: { source: string; role: ModRole; page: number; pageSize: number } }
